### PR TITLE
feat(autocomplete): add show/use completions MONGOSH-563

### PR DESCRIPTION
**Pull request by addaleax**
Add support for autocompletion for shell commands, in particular,
`show` and `use`. Autocompletion for `use` works similar to
autocompletion for `db.<collection>`.

The implementation details for use/show autocompletion here have
been left to the shell-api package, so that changes to them do not
require any autocompletion changes.

### CODE REVIEW

File: packages/cli-repl/src/mongosh-repl.ts [hunk 1]
```diff
@@ -137,11 +137,20 @@ class MongoshNodeRepl implements EvaluationListener {
         this.insideAutoComplete = true;
         try {
           // Merge the results from the repl completer and the mongosh completer.
-          const [ [replResults], [mongoshResults] ] = await Promise.all([
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          const [ [replResults], [mongoshResults, _, mongoshResultsExclusive] ] = await Promise.all([
```
Comment (mcasimir) [inline packages/cli-repl/src/mongosh-repl.ts:141]: Shouldn't `[mongoshResults,, mongoshResultsExclusive]` work too?

File: packages/autocomplete/index.ts [hunk 2]
```diff
@@ -60,6 +61,26 @@ async function completer(params: AutocompleteParameters, line: string): Promise<
   const CONFIG_COMPLETIONS = shellSignatures.ShellConfig.attributes as TypeSignatureAttributes;
   const SHARD_COMPLETE = shellSignatures.Shard.attributes as TypeSignatureAttributes;
 
+  const splitLineWhitespace = line.split(' ');
+  const command = splitLineWhitespace[0];
+  if (SHELL_COMPLETIONS[command]?.isDirectShellCommand) {
+    // If we encounter a direct shell commmand, we know that we want completions
+    // specific to that command, so we set the 'exclusive' flag on the result.
```
Comment (mcasimir) [inline packages/autocomplete/index.ts:68]: Why is this necessary? Will this also work in the browser?

File: packages/cli-repl/src/mongosh-repl.ts (hunk 1 above)
Comment (addaleax) [reply packages/cli-repl/src/mongosh-repl.ts:141]: @mcasimir Yes, I just didn’t think of that :)

File: packages/autocomplete/index.ts (hunk 2 above)
Comment (addaleax) [reply packages/autocomplete/index.ts:68]: @mcasimir Are you referring to the `'exclusive'` flag here?

The mongosh-repl.ts changes have a somewhat more specific context here, but the gist of it is that we run both the Node.js default autocompleter and our own there and combine the results; but the Node.js REPL thinks that it should complete e.g. `use a` to `use assert`, because it just looks at the `a` and finds global variables with the same name.

> Will this also work in the browser? 

I don’t think the extra array entry should hurt, we’re only using the first entry anyway:

https://github.com/mongodb-js/mongosh/blob/831220a316738c0d28f4ce464c83ba56b8b8c75d/packages/browser-runtime-core/src/autocompleter/shell-api-autocompleter.ts#L22

File: packages/cli-repl/src/mongosh-repl.ts [hunk 3]
```diff
@@ -137,11 +137,19 @@ class MongoshNodeRepl implements EvaluationListener {
         this.insideAutoComplete = true;
         try {
           // Merge the results from the repl completer and the mongosh completer.
-          const [ [replResults], [mongoshResults] ] = await Promise.all([
+          const [ [replResults], [mongoshResults,, mongoshResultsExclusive] ] = await Promise.all([
             (async() => await origReplCompleter(text) || [[]])(),
             (async() => await mongoshCompleter(text))()
           ]);
           this.bus.emit('mongosh:autocompletion-complete'); // For testing.
+
```
Comment (mcasimir) [inline packages/cli-repl/src/mongosh-repl.ts:145]: Ok, I think I've got a bit better the context. Makes sense.

File: packages/shell-api/src/database.spec.ts [hunk 4]
```diff
@@ -84,7 +84,9 @@ describe('Database', () => {
         returnType: 'AggregationCursor',
         platforms: ALL_PLATFORMS,
         topologies: ALL_TOPOLOGIES,
-        serverVersions: ALL_SERVER_VERSIONS
+        serverVersions: ALL_SERVER_VERSIONS,
+        isDirectShellCommand: false,
```
Comment (mcasimir) [inline packages/shell-api/src/database.spec.ts:88]: This is a custom "autocompleter" for arguments? And this will work with a worker thread as well cause the function will be called inside the runtime right?

Slowly getting it :).

File: packages/shell-api/src/decorators.ts [hunk 5]
```diff
@@ -142,6 +155,8 @@ export interface TypeSignature {
   deprecated?: boolean;
   returnType?: string | TypeSignature;
   attributes?: { [key: string]: TypeSignature };
+  isDirectShellCommand?: boolean;
```
Comment (mcasimir) [inline packages/shell-api/src/decorators.ts:158]: Is this used only for autocompletion or also to set up the command in the various shells? Would be great if we have to have this for it to be the source of truth about that. Hope i've understood what this means.

I think a comment here could help, the concept of "direct command" may not be immediate without context.

**Review by mcasimir — APPROVED** [review]

File: packages/shell-api/src/database.spec.ts (hunk 4 above)
Comment (addaleax) [reply packages/shell-api/src/database.spec.ts:88]: > This is a custom "autocompleter" for arguments?

The entry below is, yes.

> And this will work with a worker thread as well cause the function will be called inside the runtime right?

I *think* so, yes. The autocompleter in the browser also lives inside the same environment as e.g. the `ShellInternalState` object, and only the caller-facing `getCompletions()` method of the runtime is forwarded to the parent/rendering process, so this should work.

File: packages/shell-api/src/decorators.ts (hunk 5 above)
Comment (addaleax) [reply packages/shell-api/src/decorators.ts:158]: > Is this used only for autocompletion or also to set up the command in the various shells? Would be great if we have to have this for it to be the source of truth about that. Hope i've understood what this means.

The property on the function instance itself is what the shell evaluator uses to determine whether to allow bash-style calls to it, the property on the signature is what the autocompleter uses (like all other properties).

In either case, the source of truth is the `@directShellCommand` decorator.

> I think a comment here could help, the concept of "direct command" may not be immediate without context.

Good idea, done! :+1:

Comment (mcasimir) [reply packages/shell-api/src/decorators.ts:158]: Actually now that you called it that way .. is "bash-style" command a name that's even more clear? Of course is fine as is.

Comment (addaleax) [reply packages/shell-api/src/decorators.ts:158]: I mean, it’s not really bash-specific, it’s like that in most OS shells (even Windows cmd.exe) :)
